src/omv/ports/stm32/jpeg.c

```diff
@@ -0,0 +1,707 @@
+    } else {
+        uint8_t *new_pDataOut = pDataOut + OutDataLength;
+
+        // DMA will write data to the output buffer in __SCB_DCACHE_LINE_SIZE aligned chunks. At the
+        // end of JPEG compression the processor will manually transfer the remaining parts of the
+        // image in randomly aligned chunks. We only want to invalidate the cache of the output
+        // buffer for the initial DMA chunks. So, this code below will do that and then only
+        // invalidate aligned regions when the processor is moving the final parts of the image.
+        if (!(((uint32_t) new_pDataOut) % __SCB_DCACHE_LINE_SIZE)) {
+            SCB_InvalidateDCache_by_Addr((uint32_t *) new_pDataOut, JPEG_OUTPUT_CHUNK_SIZE);
+        }
+
+        // We are ok to receive more data.
+        HAL_JPEG_ConfigOutputBuffer(hjpeg, new_pDataOut, JPEG_OUTPUT_CHUNK_SIZE);
+    }
+}
+
+bool jpeg_compress(image_t *src, image_t *dst, int quality, bool realloc) {
+    #if (TIME_JPEG == 1)
+    mp_uint_t start = mp_hal_ticks_ms();
+    #endif
+
+    HAL_JPEG_RegisterGetDataCallback(&JPEG_state.jpeg_descr, jpeg_compress_get_data);
+    HAL_JPEG_RegisterDataReadyCallback(&JPEG_state.jpeg_descr, jpeg_compress_data_ready);
+
+    int mcu_size = 0;
+    JPEG_ConfTypeDef JPEG_Info;
+    JPEG_Info.ImageWidth = src->w;
+    JPEG_Info.ImageHeight = src->h;
+    JPEG_Info.ImageQuality = quality;
+
+    switch (src->pixfmt) {
+        case PIXFORMAT_BINARY:
+        case PIXFORMAT_GRAYSCALE:
+            mcu_size = JPEG_444_GS_MCU_SIZE;
+            JPEG_Info.ColorSpace = JPEG_GRAYSCALE_COLORSPACE;
+            JPEG_Info.ChromaSubsampling = JPEG_444_SUBSAMPLING;
+            break;
+        case PIXFORMAT_RGB565:
+        case PIXFORMAT_BAYER_ANY:
+        case PIXFORMAT_YUV_ANY:
+            mcu_size = JPEG_444_YCBCR_MCU_SIZE;
+            JPEG_Info.ColorSpace = JPEG_YCBCR_COLORSPACE;
+            JPEG_Info.ChromaSubsampling = JPEG_444_SUBSAMPLING;
+            break;
+        default:
+            break;
+    }
+
+    if (memcmp(&JPEG_state.jpeg_descr.Conf, &JPEG_Info, sizeof(JPEG_ConfTypeDef))) {
+        HAL_JPEG_ConfigEncoding(&JPEG_state.jpeg_descr, &JPEG_Info);
+    }
+
+    int src_w_mcus = (src->w + JPEG_MCU_W - 1) / JPEG_MCU_W;
+    int src_w_mcus_bytes = src_w_mcus * mcu_size;
+    int src_w_mcus_bytes_2 = src_w_mcus_bytes * 2;
+
+    // If dst->data == NULL then we need to fb_alloc() space for the payload which will be fb_free()'d
+    // by the caller. We have to alloc this memory for all cases if we return from the method.
+    if (!dst->data) {
+        uint32_t avail = fb_avail();
+        uint32_t space = src_w_mcus_bytes_2 + JPEG_ALLOC_PADDING;
+
+        if (avail < space) {
+            fb_alloc_fail();
+        }
+
+        dst->size = IMLIB_IMAGE_MAX_SIZE(avail - space);
+        dst->data = fb_alloc(dst->size, FB_ALLOC_PREFER_SIZE | FB_ALLOC_CACHE_ALIGN);
+    }
+
+    if (src->is_compressed) {
+        return true;
+    }
```

Comment:
What should happen in this case ? Should we just copy src to dst ? Note dst might be fb_alloc'd just above this.

Reply:
Oh I forgot that returning true meant overflow, I guess it's fine.

Reply:
Yep. Top level code free()s.